crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -127,7 +128,37 @@ impl Extension for QueryLimitsChecker {
         let cfg = ctx
             .data::<ServiceConfig>()
             .expect("No service config provided in schema data");
-        if query.len() > cfg.limits.max_query_payload_size as usize {
+        // Document layout of the query
+        let doc = next.run(ctx, query, variables).await?;
+
+        let is_mutation = doc
+            .operations
+            .iter()
+            .any(|(_, operation)| operation.node.ty == OperationType::Mutation);
```

Comment:
I vaguely remember we prevented having multiple operations in one request. Is that still the case? If not we risk someone piggybacking a query on a mutation.

For similar reasons, we may want to protect ourselves from someone running a mutation that is itself small (the transaction + signatures don't take up much space), but the query associated with it takes up all the space: Can we add an additional check that the total query size, less transaction bytes and signature bytes is still within the query payload limit?

Reply:
(That check may be a little tricky to pull off because it will require parsing the operation enough to get its parameters, but it seems worthwhile).

Reply:
Only one operation can be defined in a request, so one of the following: query, mutation.

Reply:
We'd have to break down the query to remove the tx_bytes + signatures, and then check the rest. I think it's possible, I'll try a stab at it.